src/cluster_slot.c

```diff
@@ -0,0 +1,207 @@
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlots() {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (server.cluster->slots[slot]) count++;
+    }
+    return count;
+}
+
+static int slotStatEntryAscCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_a.stat - entry_b.stat;
+}
+
+static int slotStatEntryDescCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_b.stat - entry_a.stat;
+}
+
+static int checkSlotAssignment(client *c, unsigned char *slots, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (server.cluster->slots[slot]) {
+            if (slots[slot]++ == 1) {
+                addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
+                return C_ERR;
+            }
+        }
+    }
+    return C_OK;
+}
+
+void addReplySlotStats(client *c, unsigned char *slots) {
+    int count = countValidSlots(slots);
+    addReplyMapLen(c, count);
+
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (!slots[slot]) continue;
+        addReplySingleSlotStat(c, slot);
+    }
+}
+
+int checkSlotAssignmentsOrReply(client *c, unsigned char *slots, int del, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (del && server.cluster->slots[slot] == NULL) {
+            addReplyErrorFormat(c,"Slot %d is already unassigned", slot);
+            return C_ERR;
+        } else if (!del && server.cluster->slots[slot]) {
+            addReplyErrorFormat(c,"Slot %d is already busy", slot);
+            return C_ERR;
+        }
+        if (slots[slot]++ == 1) {
+            addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
+            return C_ERR;
+        }
+    }
+    return C_OK;
+}
+
+int checkSlotStatsOrderByArgumentOrReply(client *c, int *order_by, int *limit, int *desc) {
+    *limit = CLUSTER_SLOTS;
+    *desc = 1;
+    *order_by = ORDER_BY_INVALID;
+
+    if (!strcasecmp(c->argv[3]->ptr, "key_count")) {
+        *order_by = ORDER_BY_KEY_COUNT;
+    }
+    if (*order_by == ORDER_BY_INVALID) {
+        addReplyError(c, "unrecognized sort column for ORDER BY. The supported columns are, 1) key_count.");
+        return C_ERR;
+    }
+
+    int i = 4; /* Next argument index, following ORDERBY */
+    while(i < c->argc) {
+        int moreargs = c->argc > i+1;
+        if (!strcasecmp(c->argv[i]->ptr,"limit") && moreargs) {
+            long tmp;
+            if (getRangeLongFromObjectOrReply(
+                c, c->argv[i+1], 1, CLUSTER_SLOTS+1, &tmp, 
```

Comment:
`CLUSTER_SLOTS+1` --> `CLUSTER_SLOTS`? Currently, something like `cluster slot-stats orderby key_count limit 16385` is (incorrectly) accepted.

Also, is the `long tmp` necessary here? Maybe something like `c, c->argv[i+1], 1, CLUSTER_SLOTS, (long*)limit,`?